### ISSUE TITLE
Access to requests in Repeater tabs

### ISSUE THREAD

**Issue by takeshixx** [PortSwigger/burp-extensions-montoya-api#10]
Currently, the site map and proxy history are accessible via:

- [history()](https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/proxy/Proxy.html#history()) in Proxy
- [requestResponses()](https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/sitemap/SiteMap.html#requestResponses()) in SiteMap

It would be nice to have something similar that allows to access requests/responses in all Repeater tabs. AFAIK to only way to extract them at the moment is to manually add each requests in Repeater to the site map.

Additionally, something similar for Intruder would be great.

**Comment by SeanBurnsUK** [comment]
Thanks for the feedback.

We do have plans to add functionality that exposes data in the burp tools. So keep an eye out for that.

**Comment by nollium** [comment]
I guess one way you could do this right now would be adding a disabled ExtensionProvidedEditor and storing all the request being passed to [isEnabledFor](https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/ui/editor/extension/ExtensionProvidedEditor.html#isEnabledFor(burp.api.montoya.http.message.HttpRequestResponse))([HttpRequestResponse](https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/http/message/HttpRequestResponse.html) requestResponse) ?

**Comment by nollium** [comment]
No, my solution should catch all [HttpRequestResponse](https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/http/message/HttpRequestResponse.html) that are opened in an [editor](ExtensionProvidedEditor), the isEnabledFor method is used by ExtensionProvidedEditor-s to chose when to display the extension's editor, and it will be called when the editor constructs (i.e., at startup or tab creation)

You are thinking of a [Proxy intercept](https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/proxy/http/ProxyRequestHandler.html), which operates at the Proxy level while my trick operates at the editor level.

Obviously this is super hacky and a real feature would be best.

**Comment by rahulr311295** [comment]
```java
String invocationType = event.invocationType().toString();
   if (invocationType == "MESSAGE_EDITOR_REQUEST") {
                    api.logging().raiseInfoEvent("Invocation Type: " + invocationType);
                    event.messageEditorRequestResponse().ifPresent(requestResponse -> {
                    HttpRequestResponse httpRequestResponse = requestResponse.requestResponse();
                    HttpRequest selectedRequest = httpRequestResponse.request();
                    HttpResponse selectedResponse = httpRequestResponse.response();
                    Main.this.getDatafromRepeater(selectedRequest, selectedResponse);
```

Here is a cheapfix i did hope it helps
[burp.api.montoya.ui.contextmenu](https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/ui/contextmenu/package-summary.html)

**Comment by takeshixx** [comment]
@jamoski3112 thanks for the example. But that would only work while the event happens, I assume they are not stored in the project files, right? I'm looking for a way to extract all requests (including Repeater) from a project file.

**Comment by NetinIn** [comment]
Hello

Could you please add more feature for Repeater. For example: 
- View a group of tabs
- View a specific tab by name
- etc.

It would be very convenient to be able to use with an MCP server.

Thanks

**Comment by Michelle-PortSwigger** [comment]
Could you give us a few examples of the types of things you'd want to do with this functionality/extensions you'd want to create to help make sure we're understanding your requirements properly? For example, what details would you want to view?

**Comment by takeshixx** [comment]
@Michelle-PortSwigger I want to do retrospective analysis of requests and responses in a Burp Suite project. The goal is to understand if a given list of API operations was covered during a pentest. To get the full testing coverage we also need to check requests sent in Repeater, just using the Proxy or SiteMap history would not include all requests.

Ideally I'd like to:
- Load a project in headless mode
- My extension extracts all requests
  - Proxy History
  - SiteMap
  - Get all repeater tabs and read ALL requests from each tab (it is important to load all requests and responses from each tab, not just the one that is visible)
  - (not sure if we can extract Intruder requests as well?)
- Analyze if those requests cover a list of API operations and inputs

**Comment by Michelle-PortSwigger** [comment]
@takeshixx Sorry for the delay in getting back to you. I need to have a chat with the team to discuss this in more detail. Before I do that can I double-check, when you say load all requests and responses are you picturing grabbing the whole request/response or just certain aspects of it such as the host, method and path?

**Comment by pS3ud0RAnD0m** [comment]
@Michelle-PortSwigger, I would like to be able to access the full req/res pairs in Repeater. Similar to how I can access them for Proxy, Proxy History, Findings, and Sitemap.

**Comment by pS3ud0RAnD0m** [comment]
Following up here with what I ended up doing with my Burp Exporter extension.

Credit to @nollium for the `ExtensionProvidedEditor` direction. I used that general idea as part of a workaround, but had to extend it pretty heavily to make it useful for both live Repeater traffic and retrospective Repeater-history export.

What I shipped is here:
- commit: [d15722a89eec01624b82e3946ae276fc4108b6e6](https://github.com/pS3ud0RAnD0m/attackframework-burp-exporter/commit/d15722a89eec01624b82e3946ae276fc4108b6e6)
- project API docs: [attackframework-burp-exporter Javadoc](https://ps3ud0rand0m.github.io/attackframework-burp-exporter/)
- If curious, just pull the latest release and export to OpenSearch or a file.

In practice, this now works well to export both:
- **historic Repeater data**
- **live Repeater data**

and it exports:
- full HTTP request/response pairs
- best-effort Repeater tab names
- best-effort Repeater group names

In short:
- for **Repeater history**, I do a best-effort Swing/Editor walk at startup to discover tabs Burp has open and capture the req/res pairs from those editor bindings
- for **live Repeater**, I track short-lived editor rebind observations and try to correlate those back to the traffic handler
- I also added best-effort `repeater_tab_name` / `repeater_group_name` export, plus trace logging and stats so users can tell which correlation path was used

This works well enough as a workaround, but it is definitely still a workaround. The hard part is that Montoya does not currently expose a first-class way to:
- enumerate all Repeater tabs/groups/history directly
- get stable Repeater tab/group metadata from the API
- correlate live Repeater traffic back to the exact tab/group with confidence when multiple identical concurrent tabs exist

Because of that, there are still cases where the exporter has to fall back to heuristics, or intentionally leave tab/group metadata null instead of guessing.

What would be much better in Montoya would be an official Repeater API that exposes:
- all open/saved Repeater tabs
- tab name
- group name
- current/full req/res pairs
- stable tab identity for live traffic correlation

That would let extensions solve this cleanly instead of relying on UI walking and editor callbacks.

**Comment by Michelle-PortSwigger** [comment]
Thanks for the feedback. I can't make any promises but we're collating all the details and the use cases here.

**Comment by takeshixx** [comment]
@Michelle-PortSwigger yes the full requests and responses. The short answer is: from a Burp Suite project file I want to be able to extract all requests and responses via an extension that I can run in headless mode.

**Comment by Michelle-PortSwigger** [comment]
@takeshixx Thank you :) I've added that into the details we're gathering.
